File: sdks/py/openai/cases/2-multi-step.py
"""
2-multi-step: Multi-step Conversation

Tests a multi-step conversation with conversation history using OpenAI SDK
and verifies that Sentry captures all spans for both API calls.
"""

import os
from openai import OpenAI
from test_runner import run_test_case


async def test_logic(inputs):
    """The actual test logic"""
    model = inputs["model"]
    system = inputs["system"]
    first_prompt = inputs["first_prompt"]
    second_prompt = inputs["second_prompt"]

    client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

    # First call
    first_response = client.chat.completions.create(
        model=model,
        messages=[
            {"role": "system", "content": system},
            {"role": "user", "content": first_prompt}
        ]
    )

    first_text = first_response.choices[0].message.content

    if not first_text:
        raise Exception("No output returned from OpenAI (first call)")

    if os.getenv("SENTRY_AI_TEST_VERBOSE") == "true":
        print(f"    First response: {first_text}")

    # Second call with conversation history
    second_response = client.chat.completions.create(
        model=model,
        messages=[
            {"role": "system", "content": system},
            {"role": "user", "content": first_prompt},
            {"role": "assistant", "content": first_text},
            {"role": "user", "content": second_prompt}
        ]
    )

    second_text = second_response.choices[0].message.content

    if not second_text:
        raise Exception("No output returned from OpenAI (second call)")

    if os.getenv("SENTRY_AI_TEST_VERBOSE") == "true":
        print(f"    Second response: {second_text}")

    return second_text


# Export test case functions
test_case = run_test_case("2-multi-step", test_logic)
main = test_case["main"]
assert_sentry = test_case["assert_sentry"]

File: sdks/js/openai/cases/2-multi-step.js
/**
 * 2-multi-step: Multi-step Conversation
 *
 * Tests a multi-step conversation with conversation history using OpenAI SDK
 * and verifies that Sentry captures all spans for both API calls.
 */

const { Sentry } = require("../setup");
const OpenAI = require("openai");
const { runTestCase } = require("../../_test-utils/test-runner.cjs");

async function testLogic(inputs) {
  const { model, system, first_prompt, second_prompt } = inputs;

  const client = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY,
  });

  // First call
  const firstCompletion = await client.chat.completions.create({
    model,
    messages: [
      { role: "system", content: system },
      { role: "user", content: first_prompt },
    ],
  });

  const firstText = firstCompletion.choices[0]?.message?.content;

  if (!firstText) {
    throw new Error("No completion returned from OpenAI (first call)");
  }

  if (process.env.SENTRY_AI_TEST_VERBOSE === "true") {
    console.log(`    First response: ${firstText}`);
  }

  // Second call with conversation history
  const secondCompletion = await client.chat.completions.create({
    model,
    messages: [
      { role: "system", content: system },
      { role: "user", content: first_prompt },
      { role: "assistant", content: firstText },
      { role: "user", content: second_prompt },
    ],
  });

  const secondText = secondCompletion.choices[0]?.message?.content;

  if (!secondText) {
    throw new Error("No completion returned from OpenAI (second call)");
  }

  if (process.env.SENTRY_AI_TEST_VERBOSE === "true") {
    console.log(`    Second response: ${secondText}`);
  }
}

module.exports = runTestCase("2-multi-step", testLogic, Sentry);
